RFC-0001-prsestissimo-metrics.md

```diff
@@ -0,0 +1,377 @@
+
+
+#### 
+**Memory footprint**
+
+We have in total 103 + 28 + 7= 138 metric keys defined in PrestoCPP and Velox. Here is the current split by metric type:
+
+
+<table>
+  <tr>
+   <td><strong>Metric Type</strong>
+   </td>
+   <td><strong>PrestoCPP Count</strong>
+   </td>
+   <td><strong>Velox Count</strong>
+   </td>
+   <td><strong>Total</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>COUNT
+   </td>
+   <td>3
+   </td>
+   <td>10
+   </td>
+   <td>13
+   </td>
+  </tr>
+  <tr>
+   <td>SUM
+   </td>
+   <td>19
+   </td>
+   <td>6
+   </td>
+   <td>25
+   </td>
+  </tr>
+  <tr>
+   <td>AVG
+   </td>
+   <td>84
+   </td>
+   <td>1
+   </td>
+   <td>85
+   </td>
+  </tr>
+  <tr>
+   <td>RATE
+   </td>
+   <td>0
+   </td>
+   <td>0
+   </td>
+   <td>0
+   </td>
+  </tr>
+  <tr>
+   <td>HISTOGRAM
+   </td>
+   <td>4
+   </td>
+   <td>11
+   </td>
+   <td>15
+   </td>
+  </tr>
+  <tr>
+   <td>
+   </td>
+   <td>
+   </td>
+   <td>Total
+   </td>
+   <td>138
+   </td>
+  </tr>
+</table>
+
+
+We can maintain a mapping of metric key (which can be anywhere between 30 to 50 characters) and metric value in the implementation of BaseStatsReporter class. Each of the above metric values is stored in a 8 byte integer type. At any moment we may have 138 × 8 = 1104 bytes of values data held in-memory. Also, we have 138 × ~50 = 6900 ~ 7K bytes for keys. In total we have 7K + 1104 ~ 8K bytes of memory. Thus holding metrics in memory has the least impact on Prestissimo workers.
+
+Given the above estimates, we decide to keep the stats in-memory and we can revisit this if the memory footprint grows.
+
+Note: Estimating the size of a histogram with quantiles is not straight forward, it depends on rotation time and error tolerance settings.
+
+
+### 
+**Metric timestamps**
+
+Since we overwrite metrics in our current design, it is likely that Prometheus may miss updates as it pulls metrics at regular intervals. The timestamps are assigned by the Prometheus server (this is recommended) and may not reflect the exact timestamp at which the metric was recorded. On the other hand, we can maintain a list of metric values and timestamps seen so far in Prestissimo and in the response to the Prometheus we can include these &lt;timestamp, value> pairs. Prometheus must be configured to honour these timestamps. But this approach is not recommended.
+
+Prestissimo emits metric values without timestamps. Prometheus DB will be assigning timestamps to the metric values after it pulls from the Prestissimo endpoint.
+
+
+### 
+**Serialisation Format**
+
+
+#### 
+**Serialize using custom interface**
+
+By default, we shall implement JSON and Prometheus Data Model serialisation of metrics. We shall expose a new Serialization interface that users can implement to customize serialization. Pros: In-house and no external dependencies. Cons: It could be challenging to keep it in sync with Prometheus data model versions. Adding support for histogram quantiles is not simple.
```

Comment:
> It could be challenging to keep it in sync with Prometheus data model versions

How many model versions are currently present?

Reply:
There aren't multiple version of the data model. In case, if there is a change in the way a metric is formatted, then we must modify our code to support that.